pkg/controller/job/job_controller.go

```diff
@@ -1662,6 +1662,11 @@ func activePodsForRemoval(job *batch.Job, pods []*v1.Pod, rmAtLeast int) []*v1.P
 		left = pods
 	}
 
+	// There is a possible panic here
+	// In real cases, if terminating pods > active, we can hit an out of bounds error
+	if len(left) < (rmAtLeast - len(rm)) {
```

Comment:
With this condition, the condition below (`len(rm) < rmAtLeast`) becomes redundand, because it checks `0 < (rmAtLeast - len(rm))`, which will be true if `len(left) < (rmAtLeast - len(rm))` for any `left`.

Reply:
Now I see why this prevents out of bounds, but I'm still not convinced this is what we want. Say, `rmAtLeast=3` and `left=1`, `rm` empty. Then, we would do nothing, but maybe we could remove the one left, shouldn't we?

Reply:
I have investigated this today. I think indeed, the PR fixes the panic, but there remains an issue with delayed scale down, which we could fix as part of this PR, or follow up. Let me demonstrate with the tests (I expect would pass).

```json
"scale down with many terminating; reproduces panic": {
  parallelism: 2,
  completions: 2,
  backoffLimit: 6,
  activePods: 4,
  failedPods: 0,
  terminatingPods: 3,
  podReplacementPolicy: podReplacementPolicy(batch.Failed),
  jobPodReplacementPolicy: true,
  expectedTerminating: ptr.To[int32](3),
  controllerTime: &referenceTime,
  expectedActive: 2,
  expectedDeletions: 2,
  expectedPodPatches: 2,
},
"scale down with terminating; demonstrates delayed deletions": {
  parallelism: 2,
  completions: 2,
  backoffLimit: 6,
  activePods: 4,
  failedPods: 0,
  terminatingPods: 1,
  podReplacementPolicy: podReplacementPolicy(batch.Failed),
  jobPodReplacementPolicy: true,
  expectedTerminating: ptr.To[int32](1),
  expectedActive: 2,
  expectedDeletions: 2,
  expectedPodPatches: 2,
},
```
I think the tests correspond to the realistic case of scale down, when we have `activePods:4`, and the completions and parallelism are reduced to `2`. We expect to delete pods to have 2 active pods. In this PR, even if the panic is fixed, then the second test would fail, as we would not delete any pods (their deletion is delayed until `terminatingPods=0`. 

So, I think, there are two issues in code to make this work. First, the check here should be to prevent panic, but we should delete as many as possible pods (inside `activePodsForRemoval`):

```golang
	// verify we still need to remove more by comparing `len(rm)` and `rmAtLeast`
	if toRemoveCnt := rmAtLeast - len(rm); toRemoveCnt > 0 {
		// verify `toRemoveCnt` does not exceed the number of pods in `left`
		if toRemoveCnt := integer.IntMin(toRemoveCnt, len(left)); toRemoveCnt > 0 {
			sort.Sort(controller.ActivePods(left))
			rm = append(rm, left[:toRemoveCnt]...)
		}
	}
```

Second, I think there is a bug on this line: https://github.com/kubernetes/kubernetes/blob/eafebcc9e368d6aeaab0ce5ec4fd56b94174d0c4/pkg/controller/job/job_controller.go#L1493C32-L1493C32. IIUC, it should be `rmAtLeast := active - wantActive` - there is no need to delete pods which are already terminating (unless I'm missing something, we can just wait for them to complete). 

Let me know what you think @kannon92 @alculquicondor

Reply:
If we drop that bug, then I don't think we need any of the guards.  Only reason why we are hitting the out of bounds is because we include terminating in rmAtLeast.

Reply:
Opened https://github.com/kubernetes/kubernetes/pull/121147 in case we go this route.